### DISCUSSION TITLE
How can I use unpublished Tauri changes?

### DISCUSSION THREAD

**Discussion by Beanow** [tauri-apps/tauri#5695] [Frequently Asked]
_Mirrored from https://tauri.app/v1/guides/faq/#how-can-i-use-unpublished-tauri-changes_

To use Tauri from GitHub (bleeding edge version) you need to change your `Cargo.toml` file and update your CLI and API.

<details>
  <summary>Pulling the Rust crate from source</summary>

Append this to your `Cargo.toml` file:

```toml title=Cargo.toml
[patch.crates-io]
tauri = { git = "https://github.com/tauri-apps/tauri", branch = "dev" }
tauri-build = { git = "https://github.com/tauri-apps/tauri", branch = "dev" }
```

This will force all your dependencies to use `tauri` and `tauri-build` from Git instead of crates.io.

</details>

<details>
  <summary>Using the Tauri CLI from source</summary>

If you are using the Cargo CLI, you can install it directly from GitHub:

```shell
cargo install --git https://github.com/tauri-apps/tauri --branch dev tauri-cli
```

If you are using the `@tauri-apps/cli` package, you will need to clone the repo and build it:

```shell
git clone https://github.com/tauri-apps/tauri
git checkout dev
cd tauri/tooling/cli/node
yarn
yarn build
```

To use it, run directly with node:

```shell
node /path/to/tauri/tooling/cli/node/tauri.js dev
node /path/to/tauri/tooling/cli/node/tauri.js build
```

Alternatively, you can run your app with Cargo directly:

```shell
cd src-tauri
cargo run --no-default-features # instead of tauri dev
cargo build # instead of tauri build - won't bundle your app though
```

</details>

<details>
  <summary>Using the Tauri API from source</summary>

It is recommended to also use the Tauri API package from source when using the Tauri crate from GitHub (though it might not be needed).
To build it from source, run the following script:

```shell
git clone https://github.com/tauri-apps/tauri
git checkout dev
cd tauri/tooling/api
yarn
yarn build
```

Now you can link it using yarn:

```shell
cd dist
yarn link
cd /path/to/your/project
yarn link @tauri-apps/api
```

Or you can change your `package.json` to point to the dist folder directly:

```json
{
  "dependencies": {
    "@tauri-apps/api": "/path/to/tauri/tooling/api/dist"
  }
}
```

</details>

**Comment by bangnokia** [comment]
I think when using the `dev` branch, it's precarious and easily breaks the app, maybe in that case we should use an alternative way like using a commit hash to avoid some unwanted changes
```
tauri = { git = "https://github.com/tauri-apps/tauri", rev = "comithash" }
```

**Reply by Beanow** [reply]
Sometimes pinning to a hash is useful, however remember this is a **bleeding edge version**.
It will break. It will be a pain to maintain long-term. And there's no security promises either.

| If you care about     | You should use      |
| --------------------- | ------------------- |
| Stability on updates  | **Normal releases** |
| Not breaking your app | **Normal releases** |
| Easy to maintain      | **Normal releases** |
| Security patches      | **Normal releases** |
